Implement Span status code specification changes - gRPC

pcwiese opened the pull request:
**Description:** 
This change implements the status code specification changes according to:

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#set-status

In a nutshell, Span status codes used to map directly to gRPC status codes and gRPC Spans used this feature to report status. The latest specification collapses the Span status code to UNSET (0), OK (1), ERROR(2), and gRPC Span status comes from the rpc.grpc.status_code required attribute.  For SDKs that have adopted this latest specification the result codes that we send to Azure monitor do not correspond to the actual gRPC error codes. Errors were always reported as 2 (Span status == error).

This change follows the backward compatibility guidelines outlined here:

https://github.com/open-telemetry/opentelemetry-proto/blob/master/opentelemetry/proto/trace/v1/trace.proto

So the logic for setting the result code is:
- Record as the Span status code (0, 1, 2)
- For HTTP, overwrite with the value from the http.status_code attribute, if any
- For gRPC, overwrite with the value from the rpc.grpc.status_code attribute, if any, otherwise use the value from the DeprecatedStatusCode

Additionally, I am recording the Span Status.message as a custom property for both Request and Dependency envelopes. This information was not captured before.

**Testing:**
Added unit tests to test this logic
Performed E2E tests locally

pcwiese committed:
Implement Span status code specification changes - gRPC

The status specification changed such that gRPC error codes for new senders comes via the rpc.grpc.status_code attribute. Old senders that haven't implemented this specification continue to report gRPC status via the DeprecatedStatusCode. This change adopts the new spec for new senders and maintains grPC backwards compatibility for old senders.

Additionally, I am recording the Status.message as a custom property for both Request and Dependency envelopes. This information was not captured before.

bogdandrutu commented:
@pcwiese do you have someone else from Azure who can do the initial review?

pcwiese commented:
> @pcwiese do you have someone else from Azure who can do the initial review?

Yeah, I'll find someone

youngbupark commented on exporter/azuremonitorexporter/trace_to_envelope.go:
```diff
@@ -106,6 +105,12 @@ func spanToEnvelope(
+	// Record the Span status message as a property if there
+	statusMessage := span.Status().Message()
+	if len(statusMessage) > 0 {
+		dataProperties["Status.message"] = statusMessage
```
Can we define `Status.message` as const? This property key would be good to start with lowercase like the others.

Also can we add simple unit-test for this property?

pcwiese replied:
After chatting we agreed to pass the raw Span status fields through as properties:

Status.Code => otel.status_code
Status.DeprecatedCode => otel.status_deprecatedcode
Status.Message => otel.status_description (description is what is spec'd)

pcwiese committed:
Address PR comments:

Pass the raw Span status field through as properties:
Status.Code => otel.status_code
Status.DeprecatedCode => otel.status_deprecatedcode
Status.Message => otel.status_description (description is what is spec'd)

youngbupark reviewed (approved):
LGTM.

Thanks!

pcwiese commented:
@bogdandrutu We are good to go now

pcwiese committed:
Misses a unit test change
